### ISSUE TITLE
Balanced MAE?

### ISSUE THREAD

**Issue by hdmjdp** [jiawei-ren/BalancedMSE#4]
Can L1 loss using your idea?

**Comment by jiawei-ren** [comment]
Sure. I have not tried the loss but It should be something like this:
```python
def bmc_loss_l1(pred, target, noise_scale):
    """Balanced MAE Loss (BMC)
    Args:
      pred: A float tensor of size [batch, 1].
      target: A float tensor of size [batch, 1].
      noise_scale: A float number or tensor. 
    Returns:
      loss: A float tensor. Balanced MAE Loss.
    """
    logits = - (pred - target.T).abs() / noise_scale   # logit size: [batch, batch]
    loss = F.cross_entropy(logits, torch.arange(pred.shape[0]))     # contrastive-like loss
    loss = loss * noise_scale.detach()  # optional: restore the loss scale, 'detach' when noise is learnable 

    return loss
```
And for the BNI version:
```python
def bni_loss_l1(pred, target, noise_scale, bucket_centers, bucket_weights):
    # Balanced MAE Loss (BNI)
    mae_term = F.l1_loss(pred, target, reduction='none') / noise_scale

    num_bucket = bucket_centers.shape[0]
    bucket_center = bucket_centers.unsqueeze(0).repeat(pred.shape[0], 1)
    bucket_weights = bucket_weights.unsqueeze(0).repeat(pred.shape[0], 1)

    balancing_term = - (pred.expand(-1, num_bucket) - bucket_center).abs() / noise_scale + bucket_weights.log()
    balancing_term = torch.logsumexp(balancing_term, dim=-1, keepdim=True)
    loss = mae_term + balancing_term
    loss = loss * noise_scale.detach()
    return loss.mean()
```
There should be a Balanced Huber loss as well.

**Comment by jiawei-ren** [comment]
For Balanced Huber Loss:

```python
def bmc_loss_huber(pred, target, noise_scale, delta=1.0):
    """Balanced Huber Loss (BMC)
    Args:
      pred: A float tensor of size [batch, 1].
      target: A float tensor of size [batch, 1].
      noise_scale: A float number or tensor.
      delta: A float number. Hyper-parameter for Huber Loss
    Returns:
      loss: A float tensor. Balanced MAE Loss.
    """
    logits = - F.huber_loss(pred, target.T, delta=delta, reduction='none') / noise_scale   # logit size: [batch, batch]
    loss = F.cross_entropy(logits, torch.arange(pred.shape[0]))     # contrastive-like loss
    loss = loss * noise_scale.detach()  # optional: restore the loss scale, 'detach' when noise is learnable

    return loss

def bni_loss_huber(pred, target, noise_scale, bucket_centers, bucket_weights, delta=1.0):
    # Balanced Huber Loss (BNI)
    huber_term = F.huber_loss(pred, target, delta=delta, reduction='none') / noise_scale

    num_bucket = bucket_centers.shape[0]
    bucket_center = bucket_centers.unsqueeze(0).repeat(pred.shape[0], 1)
    bucket_weights = bucket_weights.unsqueeze(0).repeat(pred.shape[0], 1)

    balancing_term = - F.huber_loss(pred.expand(-1, num_bucket), bucket_center, delta=delta, reduction='none') / noise_scale + bucket_weights.log()
    balancing_term = torch.logsumexp(balancing_term, dim=-1, keepdim=True)
    loss = huber_term + balancing_term
    loss = loss * noise_scale.detach()
    return loss.mean()
```

**Comment by hdmjdp** [comment]
MAE and MSE often used in speech systhiss, maybe the balanced loss can make this domain better

**Comment by jiawei-ren** [comment]
Hope to see this idea work in other fields!
